consensus/src/processes/pruning_proof/mod.rs

```diff
@@ -614,45 +712,312 @@ impl PruningProofManager {
+    }
+
+    fn estimated_blue_depth_at_level_0(&self, level: BlockLevel, level_depth: u64, current_dag_level: BlockLevel) -> u64 {
+        level_depth << current_dag_level.saturating_sub(level)
+    }
+
+    /// selected parent at level = the parent of the header at the level
+    /// with the highest blue_work (using score as work in this case)
+    fn find_selected_parent_header_at_level(
+        &self,
+        header: &Header,
+        level: BlockLevel,
+        relations_service: MTRelationsService<DbRelationsStore>,
+    ) -> PruningProofManagerInternalResult<Arc<Header>> {
+        // Logic of apply_proof only inserts parent entries for a header from the proof
+        // into the relations store for a level if there was GD data in the old stores for that
+        // header. To mimic that logic here, we need to filter out parents that are NOT in the relations_service
+        let parents = self
+            .parents_manager
+            .parents_at_level(header, level)
+            .iter()
+            .copied()
+            .filter(|parent| relations_service.has(*parent).unwrap())
+            .collect_vec()
+            .push_if_empty(ORIGIN);
+
+        let mut sp = SortableBlock { hash: parents[0], blue_work: self.headers_store.get_blue_score(parents[0]).unwrap_or(0).into() };
+        for parent in parents.iter().copied().skip(1) {
+            let sblock = SortableBlock {
+                hash: parent,
+                blue_work: self
+                    .headers_store
+                    .get_blue_score(parent)
+                    .unwrap_option()
+                    .ok_or(PruningProofManagerInternalError::NotEnoughHeadersToBuildProof(format!(
+                        "find_selected_parent_header_at_level (level {level}) couldn't find the header for block {parent}"
+                    )))?
+                    .into(),
+            };
+            if sblock > sp {
+                sp = sblock;
+            }
+        }
+        // TODO: For higher levels the chance of having more than two parents is very small, so it might make sense to fetch the whole header for the SortableBlock instead of blue_score (which will probably come from a compact header).
+        self.headers_store.get_header(sp.hash).unwrap_option().ok_or(PruningProofManagerInternalError::NotEnoughHeadersToBuildProof(
+            format!("find_selected_parent_header_at_level (level {level}) couldn't find the header for block {}", sp.hash,),
+        ))
+        // Ok(self.headers_store.get_header(sp.hash).unwrap_option().expect("already checked if compact header exists above"))
+    }
+
+    fn find_sufficient_root(
+        &self,
+        pp_header: &HeaderWithBlockLevel,
+        level: BlockLevel,
+        required_block: Option<Hash>,
+        temp_db: Arc<DB>,
+    ) -> PruningProofManagerInternalResult<(Arc<DbGhostdagStore>, Hash, Hash)> {
+        let relations_service = MTRelationsService::new(self.relations_stores.clone(), level);
```

Comment:
Maybe just build it in the constructor? It's easier this way to differentiate between temp stores and permanent ones

Reply:
Can you elaborate what you mean here? Do you mean relations_service is created outside of `find_sufficient_root`?

Reply:
I mean it should be part of `self`

Reply:
Fixed in [56b439](https://github.com/kaspanet/rusty-kaspa/pull/494/commits/56b4392178bb5a751752cafc461c3b9aa9155250)